server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java

```diff
@@ -0,0 +1,310 @@
+                ImmutableOpenMap.Builder<SnapshotShard, Long> newSnapshotShardSizes = null;
+                for (ObjectCursor<SnapshotShard> shard : knownSnapshotShardSizes.keys()) {
+                    if (onGoingSnapshotRecoveries.contains(shard.value) == false) {
+                        if (newSnapshotShardSizes == null) {
+                            newSnapshotShardSizes = ImmutableOpenMap.builder(knownSnapshotShardSizes);
+                        }
+                        newSnapshotShardSizes.remove(shard.value);
+                    }
+                }
+                if (newSnapshotShardSizes != null) {
+                    knownSnapshotShardSizes = newSnapshotShardSizes.build();
+                }
+            }
+
+            final int nbFetchers = Math.min(unknownShards, maxConcurrentFetches);
+            for (int i = 0; i < nbFetchers; i++) {
+                final int activeFetches = concurrentFetches.incrementAndGet();
+                if (activeFetches < maxConcurrentFetches + 1) {
+                    fetchNextSnapshotShard();
+                } else {
+                    final int value = concurrentFetches.decrementAndGet();
+                    assert value >= 0 : "Unexpected value: " + value;
+                }
+            }
+        } else {
+            synchronized (this) {
+                // information only needed on current master
+                knownSnapshotShardSizes = ImmutableOpenMap.of();
+                isMaster = false;
+            }
+        }
+    }
+
+    private void fetchNextSnapshotShard() {
+        boolean success = false;
+        try {
+            final SnapshotShard snapshotShard = queue.poll(0L, TimeUnit.MILLISECONDS);
+            if (snapshotShard != null) {
+                threadPool.executor(executorName()).execute(new AbstractRunnable() {
+                    @Override
+                    protected void doRun() {
+                        if (clusterService.state().nodes().isLocalNodeElectedMaster() == false) {
+                            logger.debug("skipping snapshot shard size retrieval for {} as node is no longer master", snapshotShard);
+                            return;
+                        }
+
+                        final RepositoriesService repositories = repositoriesService.get();
+                        assert repositories != null;
+                        final Repository repository = repositories.repository(snapshotShard.snapshot.getRepository());
+
+                        logger.debug("fetching snapshot shard size for {}", snapshotShard);
+                        final IndexShardSnapshotStatus status = repository.getShardSnapshotStatus(
+                            snapshotShard.snapshot().getSnapshotId(),
+                            snapshotShard.index(),
+                            snapshotShard.shardId()
+                        );
+
+                        final long snapshotShardSize = status.asCopy().getTotalSize();
```

Comment:
Maybe just make this a one liner with the above since we never really use `status` otherwise:

```java
                        final long snapshotShardSize = repository.getShardSnapshotStatus(
                                snapshotShard.snapshot().getSnapshotId(),
                                snapshotShard.index(),
                                snapshotShard.shardId()
                        ).asCopy().getTotalSize();
```

Reply:
Ok, I pushed 866ba36c7ec8a02b7a2ca2e69b0f5c3826691541